arrabiata/src/interpreter.rs

```diff
@@ -0,0 +1,962 @@
+//! [NUMBER_OF_COLUMNS] columns and [BIT_DECOMPOSITION_NUMBER_OF_CHUNKS] rows
+//! without lookups using the following layout:
+//!
+//! ```text
+//! | C1 | C2 | C3 | C4 | C5 | C6 | C7 | C8 | C9 | C10 | C11 | C12 | C13 | C14 | C15 | C16 | C17 |
+//! | y  | x  | b0 | b1 | b2 | b3 | b4 | b5 | b6 | b7  | b8  | b9  | b10 | b11 | b12 | b13 | b14 |
+//! | -- | -- | -- | -- | -- | -- | -- | -- | -- | --- | --- | --- | --- | --- | --- | --- | --- |
+//! | x0 | x  | .. | .. | .. | .. | .. | .. | .. | ... | ... | ... | ... | ... | ... | ... | ... |
+//! | x1 | x0 | .. | .. | .. | .. | .. | .. | .. | ... | ... | ... | ... | ... | ... | ... | ... |
+//! | x2 | x1 | .. | .. | .. | .. | .. | .. | .. | ... | ... | ... | ... | ... | ... | ... | ... |
+//! | .. | .. | .. | .. | .. | .. | .. | .. | .. | ... | ... | ... | ... | ... | ... | ... | ... |
+//! | x16| x15| .. | .. | .. | .. | .. | .. | .. | ... | ... | ... | ... | ... | ... | ... | ... |
+//! ```
+//!
+//! where:
+//! - `x` is the input value
+//! - for each row `i`, we have `x_i = x_{i - 1} << 15 - \sum_{j=0}^{14} 2^j b_{i * 15 + j}`
+//! - `b_i` is the i-th bit of the input value
+//!
+//! ## Handle the combinaison of constraints
+//!
+//! The prover will have to combine the constraints to generate the
+//! full circuit at the end. The constraints will be combined using a
+//! challenge (often called α) that will be generated in the verifier circuit by
+//! simulating the Fiat-Shamir transformation.
+//! The challenges will then be accumulated over time using the random coin used
+//! by the folding argument.
+//! The verifier circuit must be carefully implemented to ensure that all the
+//! messages that the prover would have sent before coining the random combiner
+//! for the constraints has been absorbed properly in the verifier circuit.
+//!
+//! Using this technique requires us a folding scheme that handles degree
+//! `5 + 1` constraints, as the challenge will be considered as a variable.
+//! The reader can refer to the folding library available in this monorepo for
+//! more contexts.
+//!
+//! ## Permutation argument
+//!
+//! Communication between rows must be done using a permutation argument. The
+//! argument we use will be a generalisation of the one used in the [PlonK
+//! paper](https://eprint.iacr.org/2019/953).
+//!
+//! The construction of the permutations will be done using the methods prefixed
+//! `save` and `load`. The index of the current row and the index of the
+//! time the value has been written will be used to generate the permutation on
+//! the fly.
+//!
+//! The permutation argument described in the PlonK paper is a kind of "inverse
+//! lookup" protocol, like Plookup. The polynomials are defined as follows:
+//!
+//! ```text
+//!            Can be seen as T[f(X)] = Χ
+//!          --------
+//!          |      |
+//! f'(X) = f(X) + β X + γ
+//!                      |--- Can be seen as the evaluation point.
```

Comment:
gamma cannot be seen as the evaluation point

Reply:
We can discuss offline. I can remove this (it is mostly a note from the draft PR that shouldn't come into this top-level doc).